feat: My Courses & My Programs switcher

farhan-arshad-dev opened the pull request:
### Description

[LEARNER-8892](https://2u-internal.atlassian.net/browse/LEARNER-8892)

- [x] The “Courses” tab is renamed to “Learn”
- [x] The “Programs” tab is removed from the tab bar
- [x] A dropdown indicator is visible on the “My Courses” screen, as in the Figma designs
- [x] Header font and style is as in the Figma designs
- [x] Tapping the dropdown indicator presents a menu of “My Courses” and “My Programs”, as in the Figma designs
- [x] The menu item for the current screen should be bolded, as in the Figma designs
- [x] Selecting a menu item:
- [x] Stays on the same screen if the selection the same as the current screen
- [x] Changes to the “My Courses” or “My Programs” screen, if the selection is different from the current screen
- [x] Dismisses the dropdown menu

Notes:
- There will still be more work to do in other App Nav tickets (e.g., header growing/shrinking, changing the Profile button to a tab, etc.)
- Dropdown menu should either be native, or a simple third party library. 

| Switcher| DropDown |
|--- | --- |
| ![Screenshot_20221019_141526](https://user-images.githubusercontent.com/43750646/196650412-830d5d8c-4d5f-4c14-972f-db2d515f2f93.png) | ![Screenshot_20221025_141640](https://user-images.githubusercontent.com/43750646/197735070-1f1e7a9b-fe85-492f-a7a4-e889c0251682.png) |

HamzaIsrar12 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/adapters/LearnDropDownAdapter.kt:
```diff
@@ -0,0 +1,43 @@
+        return LearnItemViewHolder(binding)
+    }
+
+    private class LearnItemViewHolder(val binding: LearnSelectionItemBinding) :
+        BaseViewHolder()
```
Please move this to the end of the class.

farhan-arshad-dev replied:
done.

HamzaIsrar12 reviewed (changes requested):
- The [figma](https://www.figma.com/file/Vaw3lfBXmOhTLKFnBedqE1/App-Navigation-Improvements?node-id=1114%3A1) file contain Header design for both [collapsed](https://www.figma.com/file/Vaw3lfBXmOhTLKFnBedqE1/App-Navigation-Improvements?node-id=1247%3A13601) and [expanded](https://www.figma.com/file/Vaw3lfBXmOhTLKFnBedqE1/App-Navigation-Improvements?node-id=1158%3A553) mode. Please mention somewhere that it needs to be handled somewhere.

- Programs Webview is not landscape compatible. Please raise this issue somewhere: 

- Analytics for My Courses and My Programs are working a bit different.
    - Before: Whenever the screen was visible after navigation we hit an event
    - After: The events are hitting only on the list item change event
    
- Please use a more generalized name for `navigateToMyCourses()` method.

farhan-arshad-dev commented on OpenEdXMobile/res/drawable/ic_drop_down.xml:
```diff
@@ -0,0 +1,9 @@
+    android:viewportHeight="8">
+  <path
+      android:pathData="M10.59,0.59L6,5.17L1.41,0.59L0,2L6,8L12,2L10.59,0.59Z"
+      android:fillColor="#00262B"/>
```
done.

HamzaIsrar12 replied:
Still needs to auto-format the code.

farhan-arshad-dev replied:
already formated.

farhan-arshad-dev commented on OpenEdXMobile/res/layout/fragment_learn.xml:
```diff
@@ -0,0 +1,48 @@
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:fontFamily="?attr/fontSemiBold"
+            android:paddingEnd="@dimen/learn_drop_down_icon_margin"
```
If we add the `android:paddingStart` margin, then the popup menu will appear slightly on the left side. so added `tools:ignore`.

farhan-arshad-dev commented on OpenEdXMobile/res/layout/fragment_learn.xml:
```diff
@@ -0,0 +1,48 @@
+        android:layout_width="wrap_content"
+        android:layout_height="@dimen/learn_selection_width"
+        android:layout_gravity="center"
+        android:layout_marginStart="@dimen/container_padding"
```
As per the story `collapsed` and `expanded` designs will be covered in other App Nav tickets.

farhan-arshad-dev commented on OpenEdXMobile/res/layout/learn_selection_item.xml:
```diff
@@ -0,0 +1,9 @@
+    android:gravity="center_vertical"
+    android:padding="@dimen/edx_half_margin"
+    android:textColor="@color/primaryDarkColor"
+    android:textSize="@dimen/edx_medium" />
```
> Text size should be 14sp or @dimen/edx_small

matched the font size with the design, cuz Figma files use `px` and android use `sp` for font sizes.

> Please add line-height as well: app:lineHeight="@dimen/widget_margin_double"

added.

> Padding is also a bit different from design but I think this will work 🤔

Figma files use `px` and android use `dp` for margin, and paddings.

> Plus the font and color are different on both these [designs](https://www.figma.com/file/Vaw3lfBXmOhTLKFnBedqE1/App-Navigation-Improvements?node-id=1114%3A1): Please coordinate with the design team.

will update the story description to match the decision with the design team.

farhan-arshad-dev commented on OpenEdXMobile/res/values/strings_2.xml:
```diff
@@ -12,4 +12,12 @@
+    <!-- Title of Learn tab on main dashboard screen -->
+    <string name="label_learn">Learn</string>
+    <!-- Title of user's course list screen -->
+    <string name="label_my_courses">My Courses</string>
```
no need to change cuz the design team approved the current implementation.

farhan-arshad-dev commented on OpenEdXMobile/src/androidTest/java/org/edx/mobile/test/feature/interactor/MyCoursesScreenInteractor.java:
```diff
@@ -12,7 +12,7 @@
 public class MyCoursesScreenInteractor {
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/LearnFragment.kt:
```diff
@@ -0,0 +1,113 @@
+class LearnFragment : OfflineSupportBaseFragment() {
+
+    private lateinit var binding: FragmentLearnBinding
+    private val SELECTED_POSITION = "selected_position"
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/LearnFragment.kt:
```diff
@@ -0,0 +1,113 @@
+    private fun updateScreen(itemPosition: Int, title: String) {
+        selectedItemPosition = itemPosition
+        val fragment: Fragment = when (itemPosition) {
+            0 -> {
+                environment.analyticsRegistry.trackScreenView(
+                    Analytics.Screens.MY_COURSES
+                )
+                MyCoursesListFragment().apply {
+                    arguments = this@LearnFragment.arguments
+                }
+            }
+            1 -> {
+                environment.analyticsRegistry.trackScreenView(
+                    Analytics.Screens.MY_PROGRAM
+                )
+                WebViewProgramFragment.newInstance(environment.config.programConfig.url)
+            }
+            else -> {
+                throw IllegalArgumentException("unKnown screen")
+            }
```
Sealed classes allow us to create instances with different types, but in our case we need to add only label res, so implementing enum approach.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/adapters/LearnDropDownAdapter.kt:
```diff
@@ -0,0 +1,43 @@
+    layoutRes: Int
+) : BaseListAdapter<String>(context, layoutRes, null) {
+
+    override fun render(tag: BaseViewHolder?, model: String?) {
```
renamed to `item`, and remove the `nullable`.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/adapters/LearnDropDownAdapter.kt:
```diff
@@ -0,0 +1,43 @@
+    override fun render(tag: BaseViewHolder?, model: String?) {
+        val holder = tag as LearnItemViewHolder
+        holder.binding.tvLearnItem.text = model
+        var typeface = ResourcesCompat.getFont(context, R.font.inter_regular)
+        if (isSelected(holder.position)) {
+            typeface = ResourcesCompat.getFont(context, R.font.inter_semi_bold)
+        }
+        holder.binding.tvLearnItem.typeface = typeface
```
updated

HamzaIsrar12 commented on OpenEdXMobile/res/layout/fragment_learn.xml:
```diff
@@ -0,0 +1,49 @@
+            android:textSize="@dimen/edx_large"
+            tools:ignore="RtlSymmetry" />
+
+        <ImageView
```
Based on the [design](https://www.figma.com/file/Vaw3lfBXmOhTLKFnBedqE1/App-Navigation-Improvements?node-id=2961%3A15184) the icon is changing. 🤔

HamzaIsrar12 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/LearnFragment.kt:
```diff
@@ -0,0 +1,132 @@
+        private const val SELECTED_POSITION = "selected_position"
+    }
+
+    enum class LearnScreenItems(val labelRes: Int) {
```
The name should be single i.e `LearnScreenItem`, no?

farhan-arshad-dev replied:
done.

HamzaIsrar12 reviewed (changes requested):
- My Course & My Programs Analytics are triggered twice on orientation change.
- My Course or My Programs Analytics are triggering when orientation is changed on Discovery Fragment. 

- Just a thought can we implement a **ViewModel** to store `selectedItemPosition` and `itemsList` ? 🤔

farhan-arshad-dev commented on OpenEdXMobile/res/layout/fragment_learn.xml:
```diff
@@ -0,0 +1,49 @@
+            android:layout_height="wrap_content"
+            android:layout_gravity="center"
+            android:src="@drawable/ic_drop_down"
+            tools:ignore="ContentDescription" />
```
I don't think need to set `id` or `contentDescription` cuz the parent layout has the id `ll_learn_selection`.

farhan-arshad-dev commented on OpenEdXMobile/res/layout/learn_selection_item.xml:
```diff
@@ -0,0 +1,13 @@
+    android:padding="@dimen/edx_half_margin"
+    app:lineHeight="@dimen/widget_margin_double"
+    tools:text="@string/app_name"
+    android:textColor="@color/primaryDarkColor"
```
already formatted.

HamzaIsrar12 replied:
Still pending. Please update it to `primaryBaseColor`.

farhan-arshad-dev commented on OpenEdXMobile/src/androidTest/java/org/edx/mobile/test/feature/interactor/AppInteractor.java:
```diff
@@ -26,8 +26,8 @@ public LandingScreenInteractor observeLandingScreen() {
         // Authenticated users should land here
-        public MyCoursesScreenInteractor observeMyCoursesScreen() {
-            return new MyCoursesScreenInteractor().observeMyCoursesScreen();
+        public LearnScreenInteractor observeMyCoursesScreen() {
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/androidTest/java/org/edx/mobile/test/feature/interactor/LearnScreenInteractor.java:
```diff
@@ -9,10 +9,10 @@
-        // Look for "My Courses" title which (we assume) is only present on the landing screen
-        onView(allOf(isInActionBar(), withText(R.string.label_my_courses))).check(matches(isCompletelyDisplayed()));
+public class LearnScreenInteractor {
+    public LearnScreenInteractor observeLearnScreen() {
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/MainTabsDashboardFragment.java:
```diff
@@ -70,34 +71,18 @@ protected boolean showTitleInTabs() {
+        items.add(new FragmentItemModel(LearnFragment.class,
+                getResources().getString(R.string.label_learn), R.drawable.ic_bookmark_border, getArguments(), new FragmentItemModel.FragmentStateListener() {
+            @Override
+            public void onFragmentSelected() {
+                EventBus.getDefault().post(new FragmentSelectionEvent());
+            }
+        }));
 
         if (environment.getConfig().getDiscoveryConfig().isDiscoveryEnabled()) {
             items.add(new FragmentItemModel(MainDiscoveryFragment.class,
                     getResources().getString(R.string.label_discovery), R.drawable.ic_search,
-                    getArguments(), () -> {
-                environment.getAnalyticsRegistry().trackScreenView(Analytics.Screens.FIND_COURSES);
-            }));
+                    getArguments(), () -> environment.getAnalyticsRegistry().trackScreenView(Analytics.Screens.FIND_COURSES)));
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/adapters/LearnDropDownAdapter.kt:
```diff
@@ -0,0 +1,44 @@
+class LearnDropDownAdapter(
+    context: Context,
+    layoutRes: Int
+) : BaseListAdapter<LearnFragment.LearnScreenItems>(context, layoutRes, null) {
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/LearnFragment.kt:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Subscribe
+    fun onEventMainThread(event: FragmentSelectionEvent) {
```
added.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/LearnFragment.kt:
```diff
@@ -0,0 +1,132 @@
+
+    private lateinit var binding: FragmentLearnBinding
+    private var items: ArrayList<LearnScreenItems> = arrayListOf()
+    private var selectedItemPosition = 0
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/LearnFragment.kt:
```diff
@@ -0,0 +1,132 @@
+            binding.llLearnSelection.setVisibility(false)
+        }
+        arguments?.let {
+            selectedItemPosition = it.getInt(SELECTED_POSITION, 0)
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/LearnFragment.kt:
```diff
@@ -0,0 +1,132 @@
+        outState.putInt(SELECTED_POSITION, selectedItemPosition)
+    }
+
+    override fun onDestroyView() {
```
done.

farhan-arshad-dev commented:
> My Course & My Programs Analytics are triggered twice on orientation change.

Fixed.

> My Course or My Programs Analytics are triggering when orientation is changed on Discovery Fragment.

Fixed.

> Just a thought can we implement a ViewModel to store selectedItemPosition and itemsList ? 🤔
We can implement it, but not valuable cuz the `ViewModel` is only communicating only with the screen, there is no other source of data.

HamzaIsrar12 reviewed (approved):
LGTM 👏🏻 

Merge this PR once the related [designs](https://www.figma.com/file/Vaw3lfBXmOhTLKFnBedqE1/App-Navigation-Improvements?node-id=1114%3A1) are updated on the Figma file.
